### PR TITLE
[WIP] [#93] Persistent storage & simple device display

### PR DESCRIPTION
**Pull request by Kyrrui**
I created a storage service which keeps you logged in even after refreshing or closing the browser. In order to easily clear your cache, I also implemented a simple version of the device list. To be added in later PR:
1. Testing for storageService
2. Listing of backup codes on device list
3. Logging out a device that was removed

### CODE REVIEW

**Comment by alexvandesande** [comment]
Very good start!

I noticed that if you make a connection request, that key is lost after a refresh, so that if you request in one device, accept it in another and go back to the original device after a refresh, that account will have been lost. We should probably store the keys (and state) as soon as you make a request too.

**Comment by Kyrrui** [comment]
> Very good start!
> 
> I noticed that if you make a connection request, that key is lost after a refresh, so that if you request in one device, accept it in another and go back to the original device after a refresh, that account will have been lost. We should probably store the keys (and state) as soon as you make a request too.

I agree, to make this possible we will need to add a function to the SDK where you can ask for the management keys of a contracts, then when the user hits the login page, I do a check for identity in localstorage AND I see if the device in localstorage is a management key using that api. Then I can store the identity as soon as the device makes a request, otherwise the user will just be logged in (although unable to do anything) on a refresh.

**Comment by alexvandesande** [comment]
Well you need to have that list anyway, in order to make the "devices accordion" work. It can be read directly from the contract, but you're right there should probably be a simple SDK call

**Comment by alexvandesande** [comment]
> the user will just be logged in (although unable to do anything) on a refresh.

This behavior is actually what happens when you delete your own key. You are visibly logged in, but can't do anything. You should maybe refresh the page after removing a key?

**Comment by Kyrrui** [comment]
> > the user will just be logged in (although unable to do anything) on a refresh.
> 
> This behavior is actually what happens when you delete your own key. You are visibly logged in, but can't do anything. You should maybe refresh the page after removing a key?

Hmm, that's odd - you should be redirected to the login screen when you delete your own key.

File: universal-login-example/src/views/DevicesModalView.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import React, { Component } from 'react';
```
Comment (alexvandesande) [inline universal-login-example/src/views/DevicesModalView.js]: I feel we are overusing this modal. This whole code can be perfectly replaced by a "confirm()" javascript call, which is not only less code but more native. The current code doesn't look well on mobile.

screen shot 2018-10-10 at 2 01 52 pm

Comment (Kyrrui) [reply universal-login-example/src/views/DevicesModalView.js]: Could you expand on this?

File: universal-login-example/src/services/StorageService.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import localforage from 'localforage';
+
+const IDENTITY_KEY = 'universal-login-identity';
+const DEVICES_KEY = 'universal-login-devices';
+
+class StorageService {
+  checkForIdentity() {
+    return localforage.getItem(IDENTITY_KEY).then(function (value) {
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/StorageService.js]: Why not async/await?

File: universal-login-example/src/services/StorageService.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  getDevices() {
+    return localforage.getItem(DEVICES_KEY).then(function (value) {
+      return value;
+    }).catch(function() {
+      return false;
+    });
+  }
+
+  async clearStorage() {
+    localforage.clear();
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/StorageService.js]: Does clear returns promise? if so it should be:
`return localforage.clear();`
which in this case is the same as:
`await localforage.clear();`

File: universal-login-example/src/services/StorageService.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    }).catch(function() {
+      return false;
+    });
+  }
+
+  getDevices() {
+    return localforage.getItem(DEVICES_KEY).then(function (value) {
+      return value;
+    }).catch(function() {
+      return false;
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/StorageService.js]: Should return an empty array. Functions should always return the same type.
If there are items - an array of items if no items - empty array.

Same is true for other functions. Should return null, undefined or throw an exception.

Returning different types makes it harder to use, test and destroys JIT compiler optimisations.

File: universal-login-example/src/services/StorageService.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        this.clearStorage();
+        return true;
+      }
+      // Find device in storage json object and delete
+      for (var i = 0; i < devices.length; i++) {
+        if (devices[i].publicKey == publicKey){
+          devices.splice(i, 1);
+          break;
+        }
+      }
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/StorageService.js]: Shorter version: `devices.filter( (device) => device.publicKey !== publicKey);`

File: universal-login-example/src/services/StorageService.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    }).catch(function() {
+      return false;
+    });
+  }
+
+  async removeDevice(publicKey) {
+    let devices = await this.getDevices();
+    if (devices) {
+      // If device removed is this device, clear identity
+      if (devices[0].publicKey == publicKey) {
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/StorageService.js]: always use `===`

File: universal-login-example/src/services/StorageService.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  async removeDevice(publicKey) {
+    let devices = await this.getDevices();
+    if (devices) {
+      // If device removed is this device, clear identity
+      if (devices[0].publicKey == publicKey) { 
+        this.clearStorage();
+        return true;
+      }
+      // Find device in storage json object and delete
+      for (var i = 0; i < devices.length; i++) {
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/StorageService.js]: `for (const device of devices)` but see below

File: universal-login-example/src/services/StorageService.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+  async removeDevice(publicKey) {
+    let devices = await this.getDevices();
+    if (devices) {
+      // If device removed is this device, clear identity
+      if (devices[0].publicKey == publicKey) { 
+        this.clearStorage();
+        return true;
+      }
+      // Find device in storage json object and delete
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/StorageService.js]: Using comments is often a bad small -> rather write your code in a shorter/more readable manner or extract a smaller method with an informative name

File: universal-login-example/src/components/ContentContainer.js [hunk 9]
```diff
@@ -33,6 +34,15 @@ class ContentContainer extends Component {
...
   }
 
+  async loadIdentity() {
+    const {storageService, identityService} = this.props.services;
+    let identity = await storageService.checkForIdentity();
+    if (identity) {
+      identityService.identity = identity;
+      this.setView('MainScreen');
+    }
+  }
```
Comment (marekkirejczyk) [inline universal-login-example/src/components/ContentContainer.js]: There should be no logic in Reacgt components whatsoever. All the logic should be in services.
In particular, IdentityService might have a StorageService as a dependency (passed in constructor during build time).
All persistence should be implemented in StorageService and covered with roundtrip tests. 
In the top of that, we have IdentityService which implements logic like getting identity from storage or from sdk and updating state based on those.

**Comment by Kyrrui** [comment]
@marekkirejczyk Cheers! I'm relatively new to React, so it's super helpful to learn things such as keeping business logic in the Services. I appreciate you bearing with some of my code hiccups along the way.

**Comment by marekkirejczyk** [comment]
I am looking forward to merging this PR. @Kyrrui Let me know if need any help :)

**Comment by Kyrrui** [comment]
Moving work to #114

**Pull request closed without merging**